src/backend/distributed/executor/executor.c

```diff
@@ -994,6 +1005,149 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
+		return false;
+	}
+
+	/* all the tasks are the same, so we only look one */
+	initialTask = (Task *) linitial(taskList);
+	if (initialTask->rowValuesLists != NIL)
+	{
+		/* found a multi-row INSERT */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * SequentialRunDistributedExecution gets a distributed execution and
+ * executes each individual task in the exection sequentially, one
+ * task at a time. See related function ShouldRunTasksSequentially()
+ * for more detail on the definition of SequentialRun.
+ */
+static void
+SequentialRunDistributedExecution(DistributedExecution *execution)
+{
+	DistributedPlan *distributedPlan = execution->plan;
+	Job *job = distributedPlan->workerJob;
+	List *taskList = job->taskList;
+
+	int unfinishedTaskCount = execution->unfinishedTaskCount;
+	int taskIndex = 0;
+
+	Assert(list_length(taskList) == unfinishedTaskCount);
+
+	for (taskIndex = 0; taskIndex < unfinishedTaskCount; ++taskIndex)
+	{
+		DistributedExecution *singleTaskExecution = palloc0(sizeof(DistributedExecution));
+
+		Task *taskToExecute = (Task *) list_nth(taskList, taskIndex);
+		DistributedPlan *singleTaskDistributedPlan =
+			CreateSingleTaskDistributedPlan(taskToExecute, distributedPlan);
+
+		/*
+		 * Use some fields as the same which are safe to and required to
+		 * share across single task executions.
+		 */
+		singleTaskExecution->raiseInterrupts = execution->raiseInterrupts;
+		singleTaskExecution->isTransaction = execution->isTransaction;
+		singleTaskExecution->executionStats = execution->executionStats;
+		singleTaskExecution->scanState = execution->scanState;
+
+		/* we only expect to use a single connection for a single task */
+		singleTaskExecution->targetPoolSize = 1;
+		singleTaskExecution->unfinishedTaskCount = 1;
+		singleTaskExecution->totalTaskCount = 1;
+
+		/* RunDistributedExecution() will set the sessions, be explict here */
+		singleTaskExecution->sessionList = NIL;
+		singleTaskExecution->waitEventSet = NULL;
+
+		/* we're going to execute a single task only */
+		singleTaskExecution->plan = singleTaskDistributedPlan;
+
+		/*
+		 * StartDistributedExecution is common for all the executions (including
+		 * the parent execution) and so we don't need to re-call it for each
+		 * individual singleTaskExecution execution here.
+		 */
+
+		/* simply call the regular execution function */
+		RunDistributedExecution(singleTaskExecution);
+
+		/* remember the total number of rows processed */
+		execution->rowsProcessed += singleTaskExecution->rowsProcessed;
+
+		FinishDistributedExecution(singleTaskExecution);
+	}
+}
+
+
+/*
+ * CreateSingleTaskDistributedPlan gets a task and a distributed plan. The
+ * function returns a newly allocated distributed plan where the new plan
+ * contains only the task that is provided to the function along with
+ * some of the necessary fields set.
+ */
+static DistributedPlan *
+CreateSingleTaskDistributedPlan(Task *task, DistributedPlan *originalPlan)
```

Comment:
If you agree with rest of my feedback for the implementation of this function, which would reduce this to 2 lines, what do you think about just inlining those two lines instead of creating a function for that? It is not clear from the name what this function does, so readers will need to read the function anyway, i.e. this function doesn't abstract much. Also, the comments aren't very helpful at explaining what the function does.